go-controller/pkg/node/node_dpu.go

```diff
@@ -13,137 +12,207 @@ import (
 		},
 		UpdateFunc: func(old, newer interface{}) {
-			pod := newer.(*kapi.Pod)
-			klog.Infof("Update for Pod: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
-			if util.PodWantsHostNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
-				retryPods.Delete(pod.UID)
+			oldPod := old.(*kapi.Pod)
+			newPod := newer.(*kapi.Pod)
+			klog.V(5).Infof("Update for Pod: %s/%s for network %s", newPod.Namespace, newPod.Name, netName)
+			v, ok := n.podNadCache.Load(newPod.UID)
+			if !ok {
+				klog.V(5).Infof("Skipping update for Pod %s/%s as it is not attached to network: %s",
+					newPod.Namespace, newPod.Name, netName)
 				return
 			}
-			_, retry := retryPods.Load(pod.UID)
-			if util.PodScheduled(pod) && retry {
-				if n.name != pod.Spec.NodeName {
-					retryPods.Delete(pod.UID)
-					return
+
+			networkMap := v.(map[string]*nettypes.NetworkSelectionElement)
+
+			servedCache := map[string]*util.DPUConnectionDetails{}
+			v, ok = n.servedPods.Load(newPod.UID)
+			if ok {
+				servedCache = v.(map[string]*util.DPUConnectionDetails)
+			}
+			for nadName := range networkMap {
+				podDesc := fmt.Sprintf("pod %s/%s for nad %s", newPod.Namespace, newPod.Name, nadName)
+				var oldDpuCD *util.DPUConnectionDetails
+				v, ok := servedCache[nadName]
+				if ok {
+					oldDpuCD = v
 				}
-				vfRepName, err := n.getVfRepName(pod)
-				if err != nil {
-					klog.Infof("Failed to get rep name, %s. retrying", err)
-					return
+				newDpuCD := n.podReadyToAddDPU(newPod, nadName)
+				if oldDpuCD == nil && newDpuCD == nil {
+					continue
 				}
-				podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, nil, isOvnUpEnabled, string(pod.UID),
-					"", nadName, netName, config.Default.MTU)
-				if err != nil {
-					return
+				if oldDpuCD != nil {
+					// VF already added, but new Pod has changed, we'd need to delete the old VF
+					if newDpuCD == nil || oldDpuCD.PfId != newDpuCD.PfId ||
+						oldDpuCD.VfId != newDpuCD.VfId || oldDpuCD.SandboxId != newDpuCD.SandboxId {
+						klog.Infof("Deleting the old VF since either kubelet issued cmdDEL or assigned a new VF or "+
+							"the sandbox id itself changed. Old connection details (%v), New connection details (%v)",
+							oldDpuCD, newDpuCD)
+						err := n.updatePodDPUConnStatusWithRetry(oldPod, nil, nadName)
+						if err != nil {
+							klog.Errorf("Failed to remove the old DPU connection status annotation for %s: %v", podDesc, err)
+						}
+						vfRepName, err := util.GetSriovnetOps().GetVfRepresentorDPU(oldDpuCD.PfId, oldDpuCD.VfId)
+						if err != nil {
+							klog.Errorf("Failed to get old VF representor for %s, dpuConnDetail %+v Representor port may have been deleted", podDesc, oldDpuCD, err)
+						} else {
+							err = n.delRepPort(oldDpuCD, vfRepName, nadName, podDesc)
+							if err != nil {
+								klog.Errorf("Failed to delete VF representor for %s: %v", podDesc, err)
+							}
+						}
+						delete(servedCache, nadName)
```

Comment:
same here. The whole body can be a function

Reply:
@bn222 I've fixed it based on your comments.